[WebGPU] https://webgpu.github.io/webgpu-samples/samples/computeBoids does not work

mwyrzykowski opened the pull request:
#### 36b04e3cb9bc703b4714b7379e48b2492b29d0db
<pre>
[WebGPU] <a href="https://webgpu.github.io/webgpu-samples/samples/computeBoids">https://webgpu.github.io/webgpu-samples/samples/computeBoids</a> does not work
<a href="https://bugs.webkit.org/show_bug.cgi?id=264719">https://bugs.webkit.org/show_bug.cgi?id=264719</a>
&lt;radar://118315907&gt;

Reviewed by Dan Glastonbury.

Timestamps query was added to computeBoids but we implemented the older version
in the specification.

The newer version is simpler, so support that one.

Pipeline statistics query feature was removed from WebGPU so remove it.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/WebGPU/GPUComputePassTimestampLocation.h: Removed.
* Source/WebCore/Modules/WebGPU/GPUComputePassTimestampLocation.idl: Removed.
* Source/WebCore/Modules/WebGPU/GPUComputePassTimestampWrites.h:
* Source/WebCore/Modules/WebGPU/GPURenderPassTimestampLocation.h: Removed.
* Source/WebCore/Modules/WebGPU/GPURenderPassTimestampLocation.idl: Removed.
* Source/WebCore/Modules/WebGPU/GPURenderPassTimestampWrites.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUAdapterImpl.cpp:
(WebCore::WebGPU::supportedFeatures):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUCommandEncoderImpl.cpp:
(WebCore::WebGPU::CommandEncoderImpl::beginRenderPass):
(WebCore::WebGPU::CommandEncoderImpl::beginComputePass):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUConvertToBackingContext.cpp:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUConvertToBackingContext.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUDeviceImpl.cpp:
(WebCore::WebGPU::DeviceImpl::createQuerySet):
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPU.serialization.in:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUComputePassTimestampLocation.h: Removed.
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUComputePassTimestampWrites.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPURenderPassTimestampLocation.h: Removed.
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPURenderPassTimestampWrites.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::validateComputePassDescriptor const):
(WebGPU::CommandEncoder::beginComputePass):
(WebGPU::CommandEncoder::validateRenderPassDescriptor const):
(WebGPU::CommandEncoder::beginRenderPass):
(WebGPU::CommandEncoder::resolveQuerySet):
* Source/WebGPU/WebGPU/ComputePassEncoder.h:
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::ComputePassEncoder):
(WebGPU::ComputePassEncoder::beginPipelineStatisticsQuery): Deleted.
(WebGPU::ComputePassEncoder::endPipelineStatisticsQuery): Deleted.
(wgpuComputePassEncoderBeginPipelineStatisticsQuery): Deleted.
(wgpuComputePassEncoderEndPipelineStatisticsQuery): Deleted.
* Source/WebGPU/WebGPU/HardwareCapabilities.mm:
(WebGPU::baseFeatures):
* Source/WebGPU/WebGPU/QuerySet.h:
* Source/WebGPU/WebGPU/QuerySet.mm:
(WebGPU::Device::createQuerySet):
(WebGPU::QuerySet::setOverrideLocation):
(WebGPU::QuerySet::encodeResolveCommands const):
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::RenderPassEncoder):
(WebGPU::RenderPassEncoder::beginPipelineStatisticsQuery): Deleted.
(WebGPU::RenderPassEncoder::endPipelineStatisticsQuery): Deleted.
(wgpuRenderPassEncoderBeginPipelineStatisticsQuery): Deleted.
(wgpuRenderPassEncoderEndPipelineStatisticsQuery): Deleted.
* Source/WebGPU/WebGPU/WebGPU.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebGPU/WebGPUComputePassTimestampWrites.h:
* Source/WebKit/Shared/WebGPU/WebGPURenderPassTimestampWrites.h:

Canonical link: <a href="https://commits.webkit.org/270694@main">https://commits.webkit.org/270694@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8d15a7af7c5f3c629d9b399749569c3fd206860b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26128 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4738 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27401 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28224 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23923 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26453 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6506 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2150 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23963 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26383 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3593 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22488 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28800 "Built successfully") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/26291 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/3201 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23449 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29501 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/23817 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/23835 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27392 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3242 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1441 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4648 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6284 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3709 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3564 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -141,12 +141,12 @@ static MTLStoreAction storeAction(WGPUStoreOp storeOp)
-    for (uint32_t i = 0; i < descriptor.timestampWriteCount; ++i) {
-        const auto& timestampWrite = descriptor.timestampWrites[i];
-        if (timestampWrite.queryIndex >= fromAPI(timestampWrite.querySet).count())
+    if (descriptor.timestampWrites) {
```
Should this `if` be combined with the one above it?

djg commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -141,12 +141,12 @@ static MTLStoreAction storeAction(WGPUStoreOp storeOp)
-        if (timestampWrite.queryIndex >= fromAPI(timestampWrite.querySet).count())
+    if (descriptor.timestampWrites) {
+        const auto& timestampWrite = *descriptor.timestampWrites;
+        if (timestampWrite.beginningOfPassWriteIndex >= fromAPI(timestampWrite.querySet).count() || timestampWrite.endOfPassWriteIndex >= fromAPI(timestampWrite.querySet).count())
```
There's a lot going on here. Should the `fromAPI(...)` be split out into a local variable?

```
auto querySetCount = fromAPI(timestampWrite.querySet).count();
if (timestampWrite.beginningOfPassWriteIndex >= querySetCount || timestampWrite.endOfPassWroteIndex >= querySetCount)
```

djg commented on Source/WebGPU/WebGPU/CommandEncoder.mm:
```diff
@@ -224,12 +206,12 @@ static MTLStoreAction storeAction(WGPUStoreOp storeOp)
-        if (timestampWrite.queryIndex >= fromAPI(timestampWrite.querySet).count())
+    if (descriptor.timestampWrites) {
+        const auto& timestampWrite = *descriptor.timestampWrites;
+        if (timestampWrite.beginningOfPassWriteIndex >= fromAPI(timestampWrite.querySet).count() || timestampWrite.endOfPassWriteIndex >= fromAPI(timestampWrite.querySet).count())
```
ditto.

djg reviewed (approved):
:shipit:

mwyrzykowski committed:
[WebGPU] https://webgpu.github.io/webgpu-samples/samples/computeBoids does not work
https://bugs.webkit.org/show_bug.cgi?id=264719
<radar://118315907>

Reviewed by Dan Glastonbury.

Timestamps query was added to computeBoids but we implemented the older version
in the specification.

The newer version is simpler, so support that one.

Pipeline statistics query feature was removed from WebGPU so remove it.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Modules/WebGPU/GPUComputePassTimestampLocation.h: Removed.
* Source/WebCore/Modules/WebGPU/GPUComputePassTimestampLocation.idl: Removed.
* Source/WebCore/Modules/WebGPU/GPUComputePassTimestampWrites.h:
* Source/WebCore/Modules/WebGPU/GPURenderPassTimestampLocation.h: Removed.
* Source/WebCore/Modules/WebGPU/GPURenderPassTimestampLocation.idl: Removed.
* Source/WebCore/Modules/WebGPU/GPURenderPassTimestampWrites.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUAdapterImpl.cpp:
(WebCore::WebGPU::supportedFeatures):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUCommandEncoderImpl.cpp:
(WebCore::WebGPU::CommandEncoderImpl::beginRenderPass):
(WebCore::WebGPU::CommandEncoderImpl::beginComputePass):
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUConvertToBackingContext.cpp:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUConvertToBackingContext.h:
* Source/WebCore/Modules/WebGPU/Implementation/WebGPUDeviceImpl.cpp:
(WebCore::WebGPU::DeviceImpl::createQuerySet):
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPU.serialization.in:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUComputePassTimestampLocation.h: Removed.
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPUComputePassTimestampWrites.h:
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPURenderPassTimestampLocation.h: Removed.
* Source/WebCore/Modules/WebGPU/InternalAPI/WebGPURenderPassTimestampWrites.h:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebGPU/WebGPU/CommandEncoder.mm:
(WebGPU::CommandEncoder::validateComputePassDescriptor const):
(WebGPU::CommandEncoder::beginComputePass):
(WebGPU::CommandEncoder::validateRenderPassDescriptor const):
(WebGPU::CommandEncoder::beginRenderPass):
(WebGPU::CommandEncoder::resolveQuerySet):
* Source/WebGPU/WebGPU/ComputePassEncoder.h:
* Source/WebGPU/WebGPU/ComputePassEncoder.mm:
(WebGPU::ComputePassEncoder::ComputePassEncoder):
(WebGPU::ComputePassEncoder::beginPipelineStatisticsQuery): Deleted.
(WebGPU::ComputePassEncoder::endPipelineStatisticsQuery): Deleted.
(wgpuComputePassEncoderBeginPipelineStatisticsQuery): Deleted.
(wgpuComputePassEncoderEndPipelineStatisticsQuery): Deleted.
* Source/WebGPU/WebGPU/HardwareCapabilities.mm:
(WebGPU::baseFeatures):
* Source/WebGPU/WebGPU/QuerySet.h:
* Source/WebGPU/WebGPU/QuerySet.mm:
(WebGPU::Device::createQuerySet):
(WebGPU::QuerySet::setOverrideLocation):
(WebGPU::QuerySet::encodeResolveCommands const):
* Source/WebGPU/WebGPU/RenderPassEncoder.h:
* Source/WebGPU/WebGPU/RenderPassEncoder.mm:
(WebGPU::RenderPassEncoder::RenderPassEncoder):
(WebGPU::RenderPassEncoder::beginPipelineStatisticsQuery): Deleted.
(WebGPU::RenderPassEncoder::endPipelineStatisticsQuery): Deleted.
(wgpuRenderPassEncoderBeginPipelineStatisticsQuery): Deleted.
(wgpuRenderPassEncoderEndPipelineStatisticsQuery): Deleted.
* Source/WebGPU/WebGPU/WebGPU.h:
* Source/WebKit/Scripts/webkit/messages.py:
(headers_for_type):
* Source/WebKit/Shared/WebGPU/WebGPUComputePassTimestampWrites.h:
* Source/WebKit/Shared/WebGPU/WebGPURenderPassTimestampWrites.h:

Canonical link: https://commits.webkit.org/270694@main
